### ISSUE TITLE
Black Rack Bug and possible fix

### ISSUE THREAD

**Issue by cunabula27** [VCVRack/Rack#782]
OS: Windows 10 / 4k res screen / Nvidia Geforce GTX 1070
Rack: 0.6 master

Cloned Rack master around 1-2 pm GMT yesterday (Sunday March 4) worked fine this morning after the overnight `make dep` antivirus debacle.

Ran `git pull origin master` this evening as I though the new Menu had been updated slightly from the version I had and on opening Rack the rack part was totally black:

![2018-03-05 1](https://user-images.githubusercontent.com/24782488/37004475-6a44d954-20c8-11e8-9486-5ff5ab4e4351.png)

Ran `make clean` in dep and Rack, `git submodule update --init --recursive` and recompiled everything. No change. 

Wiped everything and started from scratch. No change.

Checked all the commits marked as having happened in the last day, saw line 464 in commit `7ebf0d7` in src/window.cpp looked vaguely like it might be related.

Ran `git clone https://github.com/VCVRack/Rack.git` and changed that line in that file back to what it was. Recompiled and it's working again.

No idea why I'm afraid, not a coder yet, just looked worth a punt on the off chance and I have Notepad++

**Comment by AndrewBelt** [comment]
Could you place the line
```cpp
		debug("%d %d %f %f", windowWidth, windowHeight, gWindowRatio, gPixelRatio);
```
around src/window.cpp:463?

**Comment by Coirt** [comment]
Got that black screen bug earlier on the previous 0.6 rack master so its not new to the latest build. Was trying to display that Hz value, when it happened, trying to reproduce now happened a couple of times. Will post the piece of code that caused it if I can reproduce. If it was even that that caused it idk.

**Comment by cunabula27** [comment]
@AndrewBelt That gives:

[debug] 3840 2104 1.000000 3.000000

And it comes out the same with Rack running the newer or older line on 464, whether the screen is black or working as it's supposed to.

**Comment by Coirt** [comment]
Right... Was able to reproduce the black screen and it turns black when I turn the FREQ_PARAM all the way to the left on the LFO I'm working on with the transparent widget displaying Hz. (well trying to)

using:

```
float frq = params[FREQ_PARAM].value;
	frq = clamp(frq, -16.0f, 12.0f);
	// The default pitch is C4// Default Frq is 0.25F
	freqHz = -2.0f * powf(0.25f, frq);
```

Big long number outside the bounds of the display I think triggered it. This is not in the latest master but the one previous (previous to add menu favourites.)

**Comment by AndrewBelt** [comment]
@cunabula27 When you right click on the black area, does the Module Browser display, and can you create modules? Do some or all of the module components display correctly?
I'm thinking this is a problem with your high DPI ratio (3) and the maximum size of a framebuffer your GPU is able to create.

**Comment by Coirt** [comment]
https://www.youtube.com/watch?v=e9Fp0XvVq3U&feature=youtu.be

**Comment by AndrewBelt** [comment]
@Coirt Link to the source of that plugin?

**Comment by cunabula27** [comment]
There's something really screwy going on, I've got it to half work (really badly, but in an interesting way) but it's broken with line 464 as:

		gScene->box.size = Vec(windowWidth, windowHeight);

but it's absolutely fine with

		gScene->box.size = Vec(width, height).div(gPixelRatio / gWindowRatio);

Let me pull the latest version and see if there's a change.

I forced Rack to use my Nvidia GPU rather than the internal Intel one, it's got an 8GB frame buffer and it made no difference.

Will update this in a bit.

**Comment by cunabula27** [comment]
@Coirt that's a different issue I think, I don't lose the UI portion at the top of the screen an any point.

**Comment by Coirt** [comment]
I don't have the source up yet for 0.6. I'll upload it now and post once done.

**Comment by Coirt** [comment]
@AndrewBelt https://github.com/Coirt/Bark/tree/0.6dev

Built and working on windows as is let me know if anything is a miss. Rush job uploading think I got everything.

**Comment by AndrewBelt** [comment]
@Coirt Yeah, create a new issue

To make all these ratios correct, here's the data I have:

```
Linux/Mac/Windows normal DPI 2560 1440 1.000000 1.000000
Your Windows 4k screen 3840 2104 1.000000 3.000000
Mac Retina 1440 800 2.000000 2.000000
```

So it looks like the correct formula is
```cpp
		gScene->box.size = Vec(width, height).div(gPixelRatio);
```
Fixed in 4f9d735, could you verify?

**Comment by cunabula27** [comment]
Perfect, that's fixed it, thanks @AndrewBelt
